### PR TITLE
HPCC-21294 Ensure WsFILEIO handles newer Dropzone config structure

### PR DESCRIPTION
**Pull request by rpastrana**
- Utilizes existing logic to handle new/legacy dropzone structures

Signed-off-by: Rodrigo Pastrana <rodrigo.pastrana@lexisnexis.com>

## Type of change:
- [x] This change is a bug fix (non-breaking change which fixes an issue).
- [ ] This change is a new feature (non-breaking change which adds functionality).
- [ ] This change improves the code (refactor or other change that does not change the functionality)
- [ ] This change fixes warnings (the fix does not alter the functionality or the generated code)
- [ ] This change is a breaking change (fix or feature that will cause existing behavior to change).
- [ ] This change alters the query API (existing queries will have to be recompiled)

## Checklist:
- [x] My code follows the code style of this project.
  - [ ] My code does not create any new warnings from compiler, build system, or lint.
- [x] The commit message is properly formatted and free of typos.
  - [ ] The commit message title makes sense in a changelog, by itself.
  - [ ] The commit is signed.
- [ ] My change requires a change to the documentation.
  - [ ] I have updated the documentation accordingly, or...
  - [ ] I have created a JIRA ticket to update the documentation.
  - [ ] Any new interfaces or exported functions are appropriately commented.
- [x] I have read the CONTRIBUTORS document.
- [x] The change has been fully tested:
  - [ ] I have added tests to cover my changes.
  - [ ] All new and existing tests passed.
  - [ ] I have checked that this change does not introduce memory leaks.
  - [ ] I have used Valgrind or similar tools to check for potential issues.
- [x] I have given due consideration to all of the following potential concerns:
  - [ ] Scalability
  - [ ] Performance
  - [ ] Security
  - [ ] Thread-safety
  - [ ] Premature optimization
  - [ ] Existing deployed queries will not be broken
  - [ ] This change fixes the problem, not just the symptom
  - [ ] The target branch of this pull request is appropriate for such a change.
- [x] There are no similar instances of the same problem that should be addressed
  - [ ] I have addressed them here
  - [ ] I have raised JIRA issues to address them separately
- [ ] This is a user interface / front-end modification
  - [ ] I have tested my changes in multiple modern browsers
  - [ ] The component(s) render as expected

### CODE REVIEW

**Comment by rpastrana** [comment]
@wangkx please review

File: esp/services/ws_fileio/ws_fileioservice.cpp [hunk 1]
```diff
@@ -33,97 +33,74 @@ void CWsFileIOEx::init(IPropertyTree *cfg, const char *process, const char *serv
...
+        if (stricmp(targetDZNameOrAddress, "localhost")==0)
+            targetDZNameOrAddress = ".";
+
+        Owned<IConstDropZoneInfoIterator> dropZoneItr = env->getDropZoneIteratorByAddress(targetDZNameOrAddress);
+        ForEach(*dropZoneItr)
         {
-            const char *zonename = it->query().queryProp("@computer");
-            if (!strcmp(zonename, "."))
-                zonename = "localhost";
+            dropZoneInfo.set(&dropZoneItr->query());
```
Comment (wangkx) [inline esp/services/ws_fileio/ws_fileioservice.cpp]: Should call isECLWatchVisible() here?

Comment (wangkx) [reply esp/services/ws_fileio/ws_fileioservice.cpp]: Set the netAddr = targetDZNameOrAddress. Also set a flag to skip line 60 - 87.

File: esp/services/ws_fileio/ws_fileioservice.cpp [hunk 2]
```diff
@@ -33,97 +33,74 @@ void CWsFileIOEx::init(IPropertyTree *cfg, const char *process, const char *serv
...
+                IConstDropZoneServerInfo &serverElem = serverIter->query();
+                serverElem.getServer(netAddr.clear());
+                if (!netAddr.isEmpty())
+                    break;
+            }
+        }
+
+        dropZoneInfo->getDirectory(directory);
+        if (directory.length() != 0)
+        {
```
Comment (wangkx) [inline esp/services/ws_fileio/ws_fileioservice.cpp]: const char ch = getPathSepChar(directory);

File: esp/services/ws_fileio/ws_fileioservice.cpp [hunk 3]
```diff
@@ -33,97 +33,74 @@ void CWsFileIOEx::init(IPropertyTree *cfg, const char *process, const char *serv
...
-                zonename = "localhost";
+            dropZoneInfo.set(&dropZoneItr->query());
+            break;
+        }
+    }
 
-            if (zonename && *zonename)
+    if (dropZoneInfo)
+    {
+        char ch = '\\';
```
Comment (wangkx) [inline esp/services/ws_fileio/ws_fileioservice.cpp]: Call getPathSepChar() to get the ch (see my comment below).

**Review by wangkx — COMMENTED** [review]
@rpastrana it is better now. Please see my comments.

**Comment by rpastrana** [comment]
@wangkx please review latest commit

File: esp/services/ws_fileio/ws_fileioservice.cpp [hunk 4]
```diff
@@ -49,14 +49,17 @@ bool CWsFileIOEx::CheckServerAccess(const char* targetDZNameOrAddress, const cha
...
+            }
         }
     }
 
     if (dropZoneInfo)
     {
-        char ch = '\\';
         SCMStringBuffer dropZoneName, directory, computerName, computerAddress;
         dropZoneInfo->getName(dropZoneName);
         dropZoneInfo->getComputerName(computerName); //legacy structure
```
Comment (wangkx) [inline esp/services/ws_fileio/ws_fileioservice.cpp]: @rpastrana If you find the dropzone using getDropZoneIteratorByAddress(), you already know the network address. Then, you should be able to skip line 65 to 88.

**Comment by rpastrana** [comment]
@wangkx please review

File: esp/services/ws_fileio/ws_fileioservice.cpp [hunk 5]
```diff
@@ -49,46 +49,52 @@ bool CWsFileIOEx::CheckServerAccess(const char* targetDZNameOrAddress, const cha
...
-        char ch = '\\';
         SCMStringBuffer dropZoneName, directory, computerName, computerAddress;
-        dropZoneInfo->getName(dropZoneName);
-        dropZoneInfo->getComputerName(computerName); //legacy structure
-        if(computerName.length() != 0)
+        if (netAddr.isEmpty())
         {
-            Owned<IConstMachineInfo> machine = env->getMachine(computerName.str());
-            if (machine)
+            dropZoneInfo->getName(dropZoneName);
```
Comment (wangkx) [inline esp/services/ws_fileio/ws_fileioservice.cpp]: This line should be moved above line 65 since the dropZoneName may be used in line 109.

**Review by wangkx — COMMENTED** [review]
@rpastrana much better now. Just one comment.

**Comment by rpastrana** [comment]
one more please @wangkx

**Review by wangkx — APPROVED** [review]
@rpastrana looks good now.

**Comment by rpastrana** [comment]
@richardkchapman please merge

**Comment by richardkchapman** [comment]
@rpastrana According to the Jira this was broken in 6.4 - do we need to target 6.4.x for this fix?

**Comment by rpastrana** [comment]
@richardkchapman yes it should. Let me perform some tests on my end first

**Comment by rpastrana** [comment]
@richardkchapman apparently an api changed between 6.4 and 7.x
Owned<IEnvironmentFactory> factory = getEnvironmentFactory(true); <- bool param added, I'll create a separate PR for 6.4.x

**Pull request merged**